File: JavaPractice/src/FindWordLadderCountAndWord.java
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.Set;

public class FindWordLadderCountAndWord {

	static class WordNode {
		String word;
		int steps;
	
		public WordNode(String word, int steps) {
			this.word = word;
			this.steps = steps;
		}
	}
	
	
	public static int findLengthToWordFromDictionary(String beginWord, String endWord, Set<String> wordDict) {
		LinkedList<WordNode> queue = new LinkedList<WordNode>();
		queue.add(new WordNode(beginWord,1));
		
		//wordDict.add(endWord);
		
		while(!queue.isEmpty()) {
			
			WordNode top = queue.remove();
			String word = top.word;
			
			if(word.equals(endWord)) {
				return top.steps;
			}
			
			char [] arr = word.toCharArray();
			for(int i=0; i< arr.length; i++) {
				for(char c='a'; c<='z'; c++) {
					char temp = arr[i];
					if(arr[i] != c) {
						arr[i] = c;
					}
					
					String newWord = new String(arr);
					if(wordDict.contains(newWord)) {
						queue.add(new WordNode(newWord,top.steps + 1));
						wordDict.remove(newWord);
					}
					arr[i] = temp;
				}
			}
		}
		return 0;
	}
	public static void main(String[] args) {
		// TODO Auto-generated method stub
		Set<String> hs = new HashSet<>();
		
		hs.add("hot");
		hs.add("dot");
		hs.add("dog");
		hs.add("lot");
		hs.add("log");
		hs.add("cog"); // if I remove "cog" it will return 0 otherwise 5
		
		/*
		hs.add("poon");
		hs.add("plee");
		hs.add("same");
		hs.add("poie");
		hs.add("plie");
		hs.add("poin");
		hs.add("plea");
		*/
		System.out.println(findLengthToWordFromDictionary("hit","cog",hs));
		//System.out.println(findLengthToWordFromDictionary("toon","plea",hs));
	}

	
}

File: PythonPractice/src/FindWordLadderCountAndWord.py
from collections import deque
from typing import Set, Optional

class WordNode:
    def __init__(self, word: str, steps: int):
        self.word = word
        self.steps = steps

def find_length_to_word_from_dictionary(begin_word: str, end_word: str, word_dict: Set[str]) -> int:
    if not begin_word or not end_word or not word_dict:
        raise ValueError("Invalid input: None or empty values are not allowed.")

    queue = deque([WordNode(begin_word, 1)])

    while queue:
        top = queue.popleft()
        word = top.word

        if word == end_word:
            return top.steps

        word_list = list(word)
        for i in range(len(word_list)):
            original_char = word_list[i]
            for c in 'abcdefghijklmnopqrstuvwxyz':
                if word_list[i] != c:
                    word_list[i] = c
                new_word = ''.join(word_list)
                if new_word in word_dict:
                    queue.append(WordNode(new_word, top.steps + 1))
                    word_dict.remove(new_word)
            word_list[i] = original_char

    return 0

if __name__ == "__main__":
    word_set = {"hot", "dot", "dog", "lot", "log", "cog"}  # Example set of words
    print(find_length_to_word_from_dictionary("hit", "cog", word_set))  # Output should be 5
    # word_set = {"poon", "plee", "same", "poie", "plie", "poin", "plea"}
    # print(find_length_to_word_from_dictionary("toon", "plea", word_set))  # Test with different word list
